Add support for “known failures” to transcripts

sellout opened the pull request:
## Overview

This allows issues that haven’t yet been resolved to be reflected in the transcripts. This makes it easier to close tickets as soon as issues have been fixed, and avoids having to manually re-try transcripts from a ticket to see if the situation is the same.

Fixes #5350.

## Implementation notes

This adds a new transcript tag `:bug` to represent when ucm thinks a block is an error when it shouldn't, or isn't when it should.

- (no tag) stanza is expected to succeed; the transcript runner will complain if it fails
- `:error` stanza is expected to fail; the transcript runner will complain if it succeeds
- `:bug` stanza expected to succeed, but currently fails (a bug); the transcript runner will notify you if this changes and starts succeeding (bug is fixed).
- `:error :bug`  stanza is expected to fail, but currently succeeds (a bug); the transcript runner will notify you if this changes and starts failing (bug is fixed).

This PR also renames `:hide:all` to `:hide-all` to clarify that it's a single tag and not a combination of tags.

## Test coverage

This adds three new transcripts for open tickets exercising the two new tags.

## Loose ends

There are lots of open tickets that should also get transcripts added.

sellout committed:
Added support for known failures in transcripts

Fixes #5350.

sellout committed:
Add the first “known failure” transcript

Shows that #5337 is _not_ fixed.

sellout committed:
Add an `:incorrect` transcript tag

This is basically the opposite of `:failure` – it indicates that a
successful result is incorrect. The correct result may either be an
error (which will be caught by the transcript runner) or a different
result then what is currently expected (which won’t be caught by the
runner, but the appearance of `:incorrect` conveys to the programmer
that the output.md diff may be an improvement from the previous state).

sellout committed:
Add first `:incorrect` transcript

Shows that #5178 is _not_ fixed.

aryairani commented:
I wanna think a little more about what the label names should be, because I think `error` `failure` and `incorrect` are hard to distinguish; but otherwise this looks great.

sellout commented:
I agree. It might be worth doing #5214 before this.

aryairani commented:
How about:
* \<no-tag> : stanza is expected to succeed; runner will complain if it doesn't succeed
* `:error` : stanza is expected to fail, runner will complain if it doesn't error
* `:bug` : stanza is expected to succeed, but currently doesn't, due to a known ucm bug. runner should notify us if the behavior changes
* `:error :bug` : stanza is expected to error, but currently doesn't, due to a known ucm bug. runner should notify us if the behavior changes

And then also if it's idempotent, we want to preserve any unsuccessful input or unprocessed.

sellout committed:
Merge remote-tracking branch 'upstream/trunk' into transcript-known-failures

sellout committed:
Fix some outdated transcripts

These aren’t run by check.sh, so I missed them before pushing.

sellout commented:
Ok, I’ve switched to `:error`/`:bug`, which is a big improvement. Ready for final review.

aryairani commented on unison-cli/src/Unison/Codebase/Transcript/Runner.hs:
```diff
@@ -204,12 +205,23 @@ run isTest verbosity dir codebase runtime sbRuntime nRuntime ucmVersion baseURL
+            transcriptFailure
+              out
+              ( Text.pack . Pretty.toPlain terminalWidth $
+                  "The stanza above previously had an incorrect successful result, but now fails with"
```
my brain melts with "incorrect successful result".  

```suggestion
                  "The stanza above previously passed, but was marked as a `:bug` that should be `:error`. Now it fails with"
```

sellout replied:
I think we should come up with some consistent terms, e.g.:
- “succeeded” – had a non-error result, regardless of whether that was intended
- “errored” – had an error result, regardless of whether that was intended
- “passed” – had the intended result, regardless of whether it was success or error
- “failed” – had an unintended result, regardless of whether it was success or error (or just a different error than expected)

Doesn’t have to be those terms in particular, but I think the succeeded/errored pair should be complementary terms, and the passed/failed pair should also be complementary.

“Succeeded” is the only term there that I’m not particularly happy with.

In this case, the stanza already has `:error :bug` on it, and we want to suggest the removal of `:bug`. My original version doesn’t satisfy my terminology above because it says “fail” when it should say “error”. Maybe something like, “The stanza above is now erroring with […] as expected. Please remove `:bug`.”?

aryairani commented on unison-cli/src/Unison/Codebase/Transcript/Runner.hs:
```diff
@@ -204,12 +205,23 @@ run isTest verbosity dir codebase runtime sbRuntime nRuntime ucmVersion baseURL
+                    <> "\n"
+                    <> Pretty.border 2 msg
+                    <> "\n"
+                    <> "if this is the expected result, remove `:bug`, otherwise remove `:error`."
```
```suggestion
                    <> "If it is supposed to fail, you can remove `:bug` and hopefully close a corresponding bug ticket."
                    <> "Otherwise, if it is supposed to succeed, remove `:error`, leave `:bug`, and open a new ticket, since it is currently failing."
```
These are hard to write; let me know if I have it wrong.

sellout replied:
Oh, this is the tail of the same message – I missed that.

So, I don’t think my original or the new one are correct here. Removing `:error` is probably not what we want to do. I think what I was trying to capture here is something like “this is now erroring, which means we should remove `:bug` … but is the error the error that we’re actually expecting, or is it still the wrong error message?”

aryairani replied:
I opened a new thread below that doesn't split the message up into two conversations.

aryairani commented on unison-src/transcripts/errors/unison-hide-all-error.output.md:
````diff
@@ -4,7 +4,7 @@ When an expected error is not encountered in a `unison :hide:all:error` block
 and surface a helpful message.
 
-``` unison :hide:all:error
+``` unison :hide:all :error
````
Just curious, why does `:error` get a space but `:all` doesn't?

sellout replied:
Because `:hide:all` is really just one tag. It’s never been parsable as `:hide :all`. I just changed the printer to put spaces between, but didn’t modify the parsers.

Might be worth renaming to `:hide-all` (perhaps allowing `:hide:all` still for backward compat).

sellout replied:
I just added a commit with this change, I can back it out if you prefer.

aryairani replied:
I like `:hide-all`

sellout committed:
Replace `:hide:all` with `:hide-all`

The former was confusing because `:hide` and `:all` are not actually
separate tags.

aryairani commented on unison-cli/src/Unison/Codebase/Transcript/Runner.hs:
```diff
@@ -204,12 +205,23 @@ run isTest verbosity dir codebase runtime sbRuntime nRuntime ucmVersion baseURL
+            transcriptFailure
+              out
+              ( Text.pack . Pretty.toPlain terminalWidth $
+                  "The stanza above previously had an incorrect successful result, but now fails with"
+                    <> "\n"
+                    <> Pretty.border 2 msg
+                    <> "\n"
+                    <> "if this is the expected result, remove `:bug`, otherwise remove `:error`."
```
I think having four terms is too many, I wonder if we can do it with no more than two.

Yes to "is the error the error we're actually expecting?" that's where I was heading with the "open a new ticket if needed".

I'd be satisfied with this, wdyt:

```suggestion
                  "The stanza above marked with `:error :bug` is now failing with"
                    <> "\n"
                    <> Pretty.border 2 msg
                    <> "\n"
                    <> "so you can remove `:bug`, close any appropriate Github issues by using the commit message or PR description, and/or open a new issue if the error message is still not correct."

sellout replied:
Oh, yeah – mentioning closing the related issue is important! I’ll do the same on the the other `:bug` case. And yeah, this message looks good to me.

I think there’s still four terms, but different axes, and ones that match our tags better: passing/failing (`:error`) and expected/unexpected (`:bug`). So yeah, I like that direction.

sellout replied:
Ok, I changed the phrasing a bit again, so see what you think in the latest commits.

I also realized that the only tests added were for actual outstanding bugs, so I added an additional transcript that will continue testing the “positive” cases after those outstanding bugs are fixed, as well as two error tests for fixed bugs (which show the messages we’ve been fiddling with).

sellout committed:
Rephrase the “remove `:bug`” transcript messages

sellout committed:
Add tests for `:bug`

These exposed a few issues with the output formatting, so those are also
addressed here.
